executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,449 @@
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/memory"
+	"github.com/pingcap/tidb/util/ranger"
+	"go.uber.org/atomic"
+	"sync"
+)
+
+//IndexNestedLoopHashJoin is the hash join executor
+type IndexNestedLoopHashJoin struct {
+	IndexLookUpJoin
+	resultCh          chan *indexHashJoinResult
+	joinChkResourceCh []chan *chunk.Chunk
+}
+
+type indexHashJoinOuterWorker struct {
+	outerWorker
+	innerCh chan *indexHashJoinTask
+}
+
+type indexHashJoinInnerWorker struct {
+	innerWorker
+	matchedOuterPtrs  [][]byte
+	joiner            joiner
+	joinChkResourceCh chan *chunk.Chunk
+	resultCh          chan *indexHashJoinResult
+	taskCh            <-chan *indexHashJoinTask
+	wg                *sync.WaitGroup
+	finished          atomic.Value
+}
+
+type indexHashJoinResult struct {
+	chk *chunk.Chunk
+	err error
+	src chan<- *chunk.Chunk
+}
+
+type outerRowStatusFlag byte
+
+const (
+	_ outerRowStatusFlag = iota // outerRowUnmatched
+	outerRowMatched
+	outerRowHasNull
+)
+
+type indexHashJoinTask struct {
+	*lookUpJoinTask
+	outerRowStatus []outerRowStatusFlag
+	err            error
+}
+
+// Open implements the IndexNestedLoopHashJoin Executor interface.
+func (e *IndexNestedLoopHashJoin) Open(ctx context.Context) error {
+	// Be careful, very dirty hack in this line!!!
+	// IndexLookUpJoin need to rebuild executor (the dataReaderBuilder) during
+	// executing. However `executor.Next()` is lazy evaluation when the RecordSet
+	// result is drained.
+	// Lazy evaluation means the saved session context may change during executor's
+	// building and its running.
+	// A specific sequence for example:
+	//
+	// e := buildExecutor()   // txn at build time
+	// recordSet := runStmt(e)
+	// session.CommitTxn()    // txn closed
+	// recordSet.Next()
+	// e.dataReaderBuilder.Build() // txn is used again, which is already closed
+	//
+	// The trick here is `getStartTS` will cache start ts in the dataReaderBuilder,
+	// so even txn is destroyed later, the dataReaderBuilder could still use the
+	// cached start ts to construct DAG.
+	_, err := e.innerCtx.readerBuilder.getStartTS()
+	if err != nil {
+		return err
+	}
+
+	err = e.children[0].Open(ctx)
+	if err != nil {
+		return err
+	}
+	e.memTracker = memory.NewTracker(e.id, e.ctx.GetSessionVars().MemQuotaIndexLookupJoin)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+	e.innerPtrBytes = make([][]byte, 0, 8)
+	e.startWorkers(ctx)
+	return nil
+}
+
+func (e *IndexNestedLoopHashJoin) startWorkers(ctx context.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupJoinConcurrency
+	workerCtx, cancelFunc := context.WithCancel(ctx)
+	e.cancelFunc = cancelFunc
+	innerCh := make(chan *indexHashJoinTask, concurrency)
+	e.workerWg.Add(1)
+	ow := e.newOuterWorker(innerCh)
+	go util.WithRecovery(func() { ow.run(workerCtx) }, e.finishJoinWorkers)
+
+	e.resultCh = make(chan *indexHashJoinResult, concurrency+1)
```

Comment:
why +1

Reply:
If we want to keep a slot for error, we should make `2*concurrency+1` buffer and use another method to control goroutine concurrency. 
If the father executor slowly drains data, the buffer of this channel would always be full.